### PR TITLE
Refactors Acid Spray

### PR DESCRIPTION
**Pull request by spookydonut**
## About The Pull Request
Refactors Acid Spray and a few things that access xeno upgrade level directly

Also fixes a bug where warrior punch wouldn't apply stagger/slowdown because iscarbon(/datum/limb) is always false

## Why It's Good For The Game
removes some code duplication and atomises some changes needed for hive status refactor.

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 1]
```diff
@@ -167,6 +167,55 @@
...
+	var/damage = rand(30,40) + (X.upgrade * SPRAY_MOB_UPGRADE_BONUS)
+	apply_acid_spray_damage(damage, armor_block)
+	to_chat(src, "<span class='xenodanger'>\The [X] showers you in corrosive acid!</span>")
+
+/mob/living/carbon/proc/apply_acid_spray_damage(damage, armor_block)
+	apply_damage(damage, BURN, null, armor_block)
+
+/mob/living/carbon/human/apply_acid_spray_damage(damage, armor_block)
+	take_overall_damage(null, damage, null, null, null, armor_block)
+	if(!isyautja(src))
```
Comment (DominikPanic) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]: Early return on isyautja

**Comment by DominikPanic** [comment]
Acid spray no longer damages barricades.

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 2]
```diff
@@ -167,6 +167,56 @@
...
+/obj/structure/acid_spray_act(mob/living/carbon/Xenomorph/X)
+	if(!is_type_in_typecache(src, GLOB.acid_spray_hit))
+		return TRUE // normal density flag
+	health -= rand(40,60) + (X.upgrade * SPRAY_STRUCTURE_UPGRADE_BONUS)
+	update_health(TRUE)
+	return TRUE // normal density flag
+
+/obj/structure/razorwire/acid_spray_act(mob/living/carbon/Xenomorph/X)
+	. = ..()
+	. = FALSE // not normal density flag
```
Comment (DominikPanic) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]: Wouldn't returning false do the same here?

File: code/__DEFINES/mobs.dm [hunk 3]
```diff
@@ -465,6 +465,12 @@ var/list/global_mutations = list() // list of hidden mutation things
 #define WARRIOR_AGILITY_ARMOR 30
 #define XENO_DEADHUMAN_DRAG_SLOWDOWN 2
 
+#define SPIT_UPGRADE_BONUS ( max(0,upgrade) * 0.15 ) //increase damage by 15% per upgrade level; compensates for the loss of insane attack speeds.
+#define SPRAY_STRUCTURE_UPGRADE_BONUS 8
+#define SPRAY_MOB_UPGRADE_BONUS 4
+
+#define QUEEN_DEATH_LARVA_MULTIPLIER ( (upgrade+1)/6.0) ) // 83/66/50/33 for ancient/elder emp/elder queen/queen
```
Comment (Rohesie) [inline code/__DEFINES/mobs.dm]:
```suggestion
#define QUEEN_DEATH_LARVA_MULTIPLIER ( (upgrade+1) * 0.15) ) // 75/60/45/30 for ancient/elder emp/elder queen/queen
```
Alternatively `* 0.16) ) // 80/64/48/32` or `* 0.17) ) // 85/68/51/34`.

File: code/game/objects/structures.dm [hunk 4]
```diff
@@ -9,6 +9,7 @@
 
 	var/damage = 0
 	var/damage_cap = 500 //The point where things start breaking down.
+	var/health
```
Comment (Rohesie) [inline code/game/objects/structures.dm]: While it's positive to remove the snowflakiness of the barricade health system, then we'd have both `health` and `damage` for structures.
The best answer here is to emulate /tg/ and define `obj_integrity` and `max_integrity` at the `/obj` level, and slowly transition to it.

File: code/modules/mob/living/carbon/xenomorph/Powers.dm (hunk 2 above)
Comment (Rohesie) [reply code/modules/mob/living/carbon/xenomorph/Powers.dm]: This is a fair statement for future generations to see that any inflexible rule will create stupid situations. Subversive.
:+1:

Comment (spookydonut) [reply code/modules/mob/living/carbon/xenomorph/Powers.dm]: Yes this is tongue in cheek coding

File: code/game/objects/structures.dm (hunk 4 above)
Comment (spookydonut) [reply code/game/objects/structures.dm]: Yes at some point that'll get done, or a component will get written

**Pull request merged**